modules/obstacle.py

```diff
@@ -0,0 +1,108 @@
+"""
+Obstacle data structure.
+"""
+
+from . import lidar_detection
+
+
+class Obstacle:
+    """
+    Contains obstacles that represent lidar readings.
+
+    The following are valid obstacle contructors:
+
+    * Obstacle.create_circle_obstacle
+    * Obstacle.create_line_obstacle
+    * Obstacle.create_rect_obstacle
+    """
+
+    create_key = object()
+
+    @classmethod
+    def create_circle_obstacle(
+        cls, centre: lidar_detection.LidarDetection, radius: float
+    ) -> "tuple[bool, Obstacle.Circle | None]":
+        """
+        Circle obstacle contructor.
+        """
+        return True, Obstacle(cls.create_key, centre, radius)
+
+    @classmethod
+    def create_line_obstacle(
+        cls, start_point: lidar_detection.LidarDetection, end_point: lidar_detection.LidarDetection
+    ) -> "tuple[bool, Obstacle.Line | None]":
+        """
+        Line obstacle constructor.
+        """
+        return True, Obstacle.Line(cls.create_key, start_point, end_point)
+
+    @classmethod
+    def create_rect_obstacle(
+        cls,
+        top_left: lidar_detection.LidarDetection,
+        top_right: lidar_detection.LidarDetection,
+        bottom_left: lidar_detection.LidarDetection,
+        bottom_right: lidar_detection.LidarDetection,
+    ) -> "tuple[bool, Obstacle.Rect | None]":
+        """
+        Rect obstacle constructor.
+        """
+        return True, Obstacle.Rect(cls.create_key, top_left, top_right, bottom_left, bottom_right)
+
```

Comment:
check if the coordinates make sense - ie, x of the top left is less than the x of the top right depending on how the coordinates are defined